### ISSUE TITLE
Question about concurrency and centralized token store

### ISSUE THREAD

**Issue by gdelory** [box/box-typescript-sdk-gen#35]
- [x] I have checked that the [SDK documentation][sdk-docs] doesn't solve my issue.
- [x] I have checked that the [API documentation][api-docs] doesn't solve my issue.
- [x] I have searched the [Box Developer Forums][dev-forums] and my issue isn't already reported (or if it has been reported, I have attached a link to it, for reference).
- [x] I have searched [Issues in this repo][github-repo] and my issue isn't already reported.

### Description of the Issue

This is just a simple question, so I will leave the other sections empty. We have multiple processes using a centralized token store (so the users can use one app to do their OAuth flow). I'm trying to avoid the race conditions where 2 processes start an upload process at the exact same time, get the same expired token from the DB, and at the same time decide to refresh the token and store it back using the setToken of the TokenStorage (which calls the centralized store to store in our DB). This will lead to the process coming second to override the token of the first process in the DB, the first process won't be aware of that and will try to use its token in the following calls (coming in a few ms), as follow

```
[P1] initiate a call and call getToken on the token storage
[P2] initiate a call and call getToken on the token storage
[P1] gets token from getToken and start a call
[P2] gets token from getToken and start a call
[P1] gets 401 and decides to refresh
[P2] gets 401 and decides to refresh
[P1] asks the API for a new token
[P2] asks the API for a new token
[P1] gets a new token and call setToken on the TokenStorage
[P2] gets a new token and call setToken on the TokenStorage <== This override the AccessToken that P1 has
```

My questions are:

1. Will the first token, already overwritten by the second, work? Meaning even if I guess the refresh token is already superseded by the one from the second process, will the access token still work and call will work?
2. If the access token does work (question 1), in case the process takes more than one hour, when the 1st process will try to refresh, will it use the `getToken` of the TokenStorage first to get the latest refresh token stored by the 2nd process, or it will use the one it has (which should fail since superseded)?
3. If answer to 1 or 2 is _no it doesn't_ 😄 and we need to handle this manually (meaning test the token first in a mutex function that only one process can go through at a time), is there a way to know if a token is expired without making any call? The new SDK token seems to have lost the `acquiredAtMS` and `accessTokenTTLMS` in favor of `expiresIn`, which obviously doesn't change once stored in the DB, so I feel like there is no way to know without making a request. I didn't find any fonction like that in the code, I just see that it know it expired because it gets [401 when making a call](https://github.com/box/box-typescript-sdk-gen/blob/c92ae503f40020ee782a12c91ad3e4e78be5554c/src/fetch.ts#L169-L170). If we do have to make a call to validate the token in our mutex, what would be the fastest/smallest call just to make sure the token is refreshed if needed, so we can then leave the mutex? Is there some type of `ping` function doing nothing more that answer `OK`?
4. And writing 3, I actually realized that it doesn't solve the problem when the token will expire exactly at the same time on both processes if they upload for more than one hour, since the mutex check would be done only after creating the client. So do you think it would be an option to overwrite the `refreshToken` function (which I see is called [here](https://github.com/box/box-typescript-sdk-gen/blob/c92ae503f40020ee782a12c91ad3e4e78be5554c/src/fetch.ts#L171) in case of 401) to call itself, but in a mutex (we use a global lock in redis). Again, if 1 and 2 are working, then I think it's not an issue, the 2nd process will also overwrite the token of the first process, but if the access token still works, and it will request the latest refresh token from the token store one hour later when it needs to refresh, everything should work 😄 

### Steps to Reproduce

5. Click on '....'
6. Scroll down to '....'
7. See error

### Versions Used

Typescript SDK: 
Platform: 
Node.js (if applicabe): 

[sdk-docs]: https://github.com/box/box-ts-sdk-generated/tree/main/docs
[api-docs]: https://developer.box.com/docs
[dev-forums]: https://community.box.com/t5/Platform-and-Development-Forum/bd-p/DeveloperForum
[github-repo]: https://github.com/box/box-ts-sdk-generated/search?type=Issues

**Comment by gdelory** [comment]
I'll try to answer my own question a bit. I see in the code that [refreshToken does first retrieve the latest AccessToken](https://github.com/box/box-typescript-sdk-gen/blob/c92ae503f40020ee782a12c91ad3e4e78be5554c/src/oauth.ts#L169) from the TokenStorage (in our case the DB), so I think the refresh for the first process is not an issue, even if its token was overwritten a few ms after being acquired.

So I guess that if the access token is still working even if another `AccessToken` was generated for process 2, then I can answer that 1 and 2 do work, so there is no need for mutex. The whole flow will be the following:

```
[P1] initiate a call and call getToken on the token storage
[P2] initiate a call and call getToken on the token storage
[P1] gets token from getToken and start a call
[P2] gets token from getToken and start a call
[P1] gets 401 and decides to refresh
[P2] gets 401 and decides to refresh
[P1] asks the API for a new token
[P2] asks the API for a new token
[P1] gets a new token (T1) and call setToken on the TokenStorage
[P2] gets a new token (T2) and call setToken on the TokenStorage <== This override the AccessToken that P1 has
[P1] accesses content using T1.accessToken for one hour long (at this point T1.refreshToken is already invalid, but T1.accessToken still works???
[P2] accesses content using T2.accessToken for one hour long
[P1] gets 401 and decides to refresh
[P2] gets 401 and decides to refresh (this happens at the same ms than the previous line)
[P1] gets a new token (T3) and call setToken on the TokenStorage
[P2] gets a new token (T4) and call setToken on the TokenStorage <== This override the AccessToken that P1 has
[P1] accesses content using T3.accessToken for one hour long (at this point T3.refreshToken is already invalid, but T3.accessToken still works???
[P2] accesses content using T4.accessToken for one hour long
Repeat...
```

So assuming the assumption _at this point T1.refreshToken is already invalid, but T1.accessToken still works???_ is true, I think everything should work fine, even if it's 1000 processes instead of 2.

Of course P2 could be first sometimes, this is just to understand the concept.

**Comment by arjankowski** [comment]
Hello @gdelory and thank you for this interesting question!

The old `AccessToken` object, when refreshed using the refreshToken, stops working, and from that point on, you **need** to use a new token.

However, the situation is interesting when it comes to the refresh token. If we refreshed the `AccessToken` object using the refreshToken but haven't made any API requests yet using the new access token, then we can reuse the old refreshToken for refreshing, and it will return the same new AccessToken object as in the previous refresh.

So, if two processes simultaneously request and refresh the token, and there are no other API calls between them, everything will work fine because in both cases, upon refresh, you will receive the same new AccessToken object.

However, there may be a situation where P2 receives a 401 error and starts to refresh the token, but just before that, the same action is taken by process P1 and it has already updated the tokenStorage. In this case, P2 will overwrite the tokenStorage, and then P1 will once again receive a 401 error and refresh the token.

In summary, it seems that the current token problem will occur in the case of concurrency and centralized token storage. May I ask how you addressed this issue in the legacy box-node-sdk?

Regards,
Artur

**Comment by gdelory** [comment]
Hey @arjankowski , thanks so much for you answer. I have a few more questions following this, and I'll try to answer yours.

> The old AccessToken object, when refreshed using the refreshToken, stops working, and from that point on, you need to use a new token.

Can you be a bit more specific. I use `AccessToken` as the JSON object which contains both the access token and the refresh token. I understand that sometimes although the refresh token is now invalid, the access token still works. So here, when you say it stops working, are both stop working, or only the refresh token. This makes a huge difference, because if the access token still works until it is supposed to expire, then we have to race condition like I show in my flow, one hour later it will use the getToken function from the TokenStorage and will get the latest one stored by the other process. But if the access token is also invalid, then all subsequent calls from P1 will fail.

Although, maybe then it will try to refresh again, which will trigger a call to getToken, and it would get the latest token stored by P2 and be able to refresh?

I think I will have to force the race condition here with some pause between every step to see what really happens 😄 

> However, the situation is interesting when it comes to the refresh token. If we refreshed the AccessToken object using the refreshToken but haven't made any API requests yet using the new access token, then we can reuse the old refreshToken for refreshing, and it will return the same new AccessToken object as in the previous refresh.

That's interesting to know indeed that the old refresh token will still work and return the same new refresh token P2 got as well. However that would require the perfect race condition, and this I can't enforce, maybe some calls will have been done already 😉 

> However, there may be a situation where P2 receives a 401 error and starts to refresh the token, but just before that, the same action is taken by process P1 and it has already updated the tokenStorage. In this case, P2 will overwrite the tokenStorage, and then P1 will once again receive a 401 error and refresh the token.

Well to be honest if P1 knows to refresh again, that's fine for us. What we are trying to avoid is the situation where one of the process just throw an exception and doesn't do the upload, that's a problem for us because then we have unstable processes. But if it just need to go through an extra refresh and uploads fine, I'm perfectly fine with that.

> May I ask how you addressed this issue in the legacy box-node-sdk?

Sure, but TBH I'm not sure we fixed it 100% either there. What we did is that we first retrieved the token from the centralized token with a special endpoint (not box related, but basically the same as used in the getToken of the store), then because of the acquire and TTL properties we could check if it was expired, then we could a server side endpoint, which was using a mutex to make sure two processes couldn't call it at the same time (well they could but the other processes would wait to get an answer). Then we were refreshing the token on the server side and storing it in the mutexed function, then we could create a persistent client with:

```
sdk.getPersistentClient(existingToken, this.getTokenStore(userId))
```

You can see it solved the first initialization race condition, but I don't think it solved the auto refresh that could happen one hour later.

Anyway, I will play a bit with it, but again, if you say it just triggers one more refresh but nothing really fails, this is perfect.

**Comment by arjankowski** [comment]
hi @gdelory,

Referring to the `AccessToken`, I meant the returned object containing fields such as `accessToken` and `refreshToken`. Generally, at any given moment, when making an API call, only one `AccessToken.accessToken` is valid. If you have an `AccessToken` object (old), whose `AccessToken.accessToken` will soon expire, and you decide to refresh it, receiving a new `AccessToken` object (new), then the old `AccessToken.accessToken` becomes invalid, and using it for a request from now will result in a 401 error.

However, if process P1 uses the old `AccessToken.accessToken` for a long-running operation, and during the execution of this operation, process P2 refreshes the `AccessToken` object because its validity has expired, then process P1, from my observation, will continue to execute for some time, and during that time, P2 should already save the new `AccessToken` to storage.

I think that a possible solution to your problem might be using separate Box applications, where each of them will have independent token storage.

Summarizing this discussion, so far, we haven't focused on concurrency and centralized token store in this SDK, so it might be challenging for us to propose something in this case. Rather, we have been concentrating within our SDK on a single process without sharing storage.


But please do let us know your conclusions

Regards,
Artur

**Comment by uncvrd** [comment]
I guess I just want to share my (naive) approach to prevent two clients attempting to refresh a token at the same time was to use a cache (redis/nats/in-memory) to "lock" the client when there are multiple clients attempting to connect for the same user (in my use-case a workspace). Here's my static method that initiates a class. 

The `getBoxAccessToken` and `storeAccessToken` are the same methods I use for the `get` and `store` methods that I use in my `TokenStorage` implementation

Please let me know if anyone else has better implementations. I was receiving a lot of `bad refresh_token` errors since I had multiple clients trying to create a new access token. After implementing, I havent seen this come through

```
static async initialize(workspaceId?: number) {
        if (!!workspaceId) {
            const accessToken = await getBoxAccessToken(workspaceId);

            if (accessToken?.refreshToken) {
                const cacheKey = JSON.stringify({
                    type: "box-api-lock",
                    workspaceId,
                });

                let isLocked: boolean = cache.get<boolean>(cacheKey) ?? false;

                if (isLocked) {
                    let attempts = 1;
                    const maxAttempts = 20;

                    do {

                        attempts++;

                        await sleep(500);

                        isLocked = cache.get<boolean>(cacheKey) ?? false;
                    } while (isLocked && attempts < maxAttempts);
                } else {
                    try {
                        cache.set(cacheKey, true);

                        const response = await fetch("https://api.box.com/oauth2/token", {
                            method: "POST",
                            headers: {
                                "Content-Type": "application/x-www-form-urlencoded",
                            },
                            body: new URLSearchParams({
                                client_id: env.BOX_CLIENT_ID,
                                client_secret: env.BOX_CLIENT_SECRET,
                                refresh_token: accessToken.refreshToken,
                                grant_type: "refresh_token",
                            }),
                        });

                        const body = (await response.json()) as RawRefreshTokenResponse;

                        await storeAccessToken(workspaceId, {
                            accessToken: body.access_token,
                            expiresIn: body.expires_in,
                            issuedTokenType: body.issued_token_type,
                            refreshToken: body.refresh_token,
                            restrictedTo: body.restricted_to,
                            tokenType: body.token_type,
                        });
                    } catch (error) {
                        if (error instanceof Error) {
                            console.log(error.message);
                        }
                    } finally {
                        cache.del(cacheKey);
                    }
                }
            }

            const auth = new BoxOAuth({
                config: new OAuthConfig({
                    clientId: env.BOX_CLIENT_ID,
                    clientSecret: env.BOX_CLIENT_SECRET,
                    tokenStorage: new CustomStorage(workspaceId),
                }),
            });

            const client = new BoxClient({
                auth,
            });

            return new BoxUserAPI(auth, client);
        } else {
            const auth = new BoxOAuth({
                config: new OAuthConfig({
                    clientId: env.BOX_CLIENT_ID,
                    clientSecret: env.BOX_CLIENT_SECRET,
                }),
            });

            const client = new BoxClient({
                auth,
            });

            return new BoxServerAPI(auth, client);
        }
    }
```
